File: packages/schema-domain/src/activity-memory-graph.ts
import {
  type Infer,
  NonEmptyStringSchema,
  Schema,
  withParser,
} from './effect';

const NonNegativeCountSchema = Schema.Number.pipe(Schema.nonNegative(), Schema.int());
const NullableTextSchema = Schema.Union(NonEmptyStringSchema, Schema.Null);

const ActivityMemoryGraphEvidenceReferenceSchema = Schema.Struct({
  evidenceReferenceId: NonEmptyStringSchema,
  kind: NonEmptyStringSchema,
  observedAt: NonEmptyStringSchema,
});

const ActivityMemoryGraphParentActionReferenceSchema = Schema.Struct({
  actionReferenceId: NonEmptyStringSchema,
  actor: Schema.Struct({
    actorId: NonEmptyStringSchema,
    role: NonEmptyStringSchema,
  }),
  policyVersion: NonEmptyStringSchema,
  createdAt: NonEmptyStringSchema,
});

const ActivityMemoryGraphDeviceReferenceSchema = Schema.Struct({
  deviceId: NonEmptyStringSchema,
  childProfileId: NullableTextSchema,
  label: NonEmptyStringSchema,
  platform: NonEmptyStringSchema,
});

const ActivityMemoryGraphChildProfileReferenceSchema = Schema.Struct({
  childProfileId: NonEmptyStringSchema,
  displayName: NonEmptyStringSchema,
});

export const ActivityMemoryGraphEntryStatusSchema = withParser(
  Schema.Literal('usable', 'degraded', 'stale', 'rejected')
);

export const ActivityMemoryGraphNodeKindSchema = withParser(
  Schema.Literal(
    'child-profile',
    'device',
    'browser-url',
    'domain',
    'video',
    'app',
    'game',
    'activity-session'
  )
);

export const ActivityMemoryGraphEdgeKindSchema = withParser(
  Schema.Literal(
    'visited',
    'watched',
    'played',
    'active-during',
    'performed-by-child',
    'derived-from-evidence'
  )
);

export const ActivityMemoryGraphQueryKindSchema = withParser(
  Schema.Literal(
    'visited-urls',
    'played-games',
    'watched-videos',
    'activity-by-time-range',
    'explain-evidence'
  )
);

export const ActivityMemoryGraphTraceSchema = withParser(
  Schema.Struct({
    entryStatus: ActivityMemoryGraphEntryStatusSchema,
    sourceEvidenceReferences: Schema.Array(ActivityMemoryGraphEvidenceReferenceSchema),
    sourcePolicyVersion: NullableTextSchema,
    sourceParentActionReferences: Schema.Array(ActivityMemoryGraphParentActionReferenceSchema),
    generatedAt: NonEmptyStringSchema,
    expiresAt: NullableTextSchema,
    confidence: Schema.Number.pipe(Schema.nonNegative()),
    derivedIndexVersion: NonEmptyStringSchema,
    degradedReasons: Schema.Array(NonEmptyStringSchema),
  }).pipe(
    Schema.filter(
      (trace) =>
        trace.sourceEvidenceReferences.length > 0 ||
        trace.sourcePolicyVersion !== null ||
        trace.sourceParentActionReferences.length > 0 ||
        'Expected activity memory graph trace to cite source context'
    )
  )
);

export const ActivityMemoryGraphNodeSchema = withParser(
  Schema.Struct({
    graphId: NonEmptyStringSchema,
    nodeId: NonEmptyStringSchema,
    nodeKind: ActivityMemoryGraphNodeKindSchema,
    label: NonEmptyStringSchema,
    childProfile: Schema.Union(ActivityMemoryGraphChildProfileReferenceSchema, Schema.Null),
    device: Schema.Union(ActivityMemoryGraphDeviceReferenceSchema, Schema.Null),
    trace: ActivityMemoryGraphTraceSchema,
  })
);

export const ActivityMemoryGraphEdgeSchema = withParser(
  Schema.Struct({
    graphId: NonEmptyStringSchema,
    edgeId: NonEmptyStringSchema,
    edgeKind: ActivityMemoryGraphEdgeKindSchema,
    fromNodeId: NonEmptyStringSchema,
    toNodeId: NonEmptyStringSchema,
    observedFrom: NonEmptyStringSchema,
    observedUntil: NullableTextSchema,
    durationMs: Schema.Union(NonNegativeCountSchema, Schema.Null),
    trace: ActivityMemoryGraphTraceSchema,
  })
);

export const ActivityMemoryGraphReadModelSchema = withParser(
  Schema.Struct({
    schemaVersion: NonNegativeCountSchema,
    generatedAt: NonEmptyStringSchema,
    custody: NonEmptyStringSchema,
    capabilityStatus: NonEmptyStringSchema,
    query: Schema.Struct({
      queryId: NonEmptyStringSchema,
      queryKind: ActivityMemoryGraphQueryKindSchema,
      childProfile: Schema.Union(ActivityMemoryGraphChildProfileReferenceSchema, Schema.Null),
      device: ActivityMemoryGraphDeviceReferenceSchema,
      timeRange: Schema.Struct({
        observedFrom: NonEmptyStringSchema,
        observedUntil: NonEmptyStringSchema,
      }),
      asOf: NonEmptyStringSchema,
      limit: NonNegativeCountSchema,
    }),
    readAt: NonEmptyStringSchema,
    nodes: Schema.Array(ActivityMemoryGraphNodeSchema),
    edges: Schema.Array(ActivityMemoryGraphEdgeSchema),
    returnedNodeCount: NonNegativeCountSchema,
    returnedEdgeCount: NonNegativeCountSchema,
    omittedEdgeCount: NonNegativeCountSchema,
    degradedReasons: Schema.Array(NonEmptyStringSchema),
  }).pipe(
    Schema.filter(
      (model) => model.returnedNodeCount === model.nodes.length || 'Expected memory graph node count match'
    ),
    Schema.filter(
      (model) => model.returnedEdgeCount === model.edges.length || 'Expected memory graph edge count match'
    )
  )
);

export type ActivityMemoryGraphReadModel = Infer<typeof ActivityMemoryGraphReadModelSchema>;
export type ActivityMemoryGraphNode = ActivityMemoryGraphReadModel['nodes'][number];
export type ActivityMemoryGraphEdge = ActivityMemoryGraphReadModel['edges'][number];
export type ActivityMemoryGraphNodeId = ActivityMemoryGraphNode['nodeId'];

export function parseActivityMemoryGraphDigest(digest: string): ActivityMemoryGraphReadModel | null {
  const decoded = parseDigest(digest);
  if (decoded === null) {
    return null;
  }
  const parsed = ActivityMemoryGraphReadModelSchema.safeParse(decoded);
  if (!parsed.success) {
    return null;
  }
  return parsed.data;
}

function parseDigest(digest: string): unknown {
  try {
    return JSON.parse(digest) as unknown;
  } catch {
    return null;
  }
}

File: crates/agent-protocol/src/activity_memory_graph.rs
use serde::{Deserialize, Serialize};

use crate::{
    activity::policy::ParentEvidenceReference,
    activity::policy_context::{ChildProfileReference, ParentDeviceReference},
    enforcement::ParentActionReference,
};

pub const ACTIVITY_MEMORY_GRAPH_SCHEMA_VERSION: u16 = crate::ACTIVITY_MEMORY_GRAPH_SCHEMA_VERSION;
pub const ACTIVITY_MEMORY_GRAPH_CUSTODY_ACTIVITY_STORE: &str = "child-device-activity-store";
pub const ACTIVITY_MEMORY_GRAPH_CAPABILITY_READY: &str = "ready";
pub const ACTIVITY_MEMORY_GRAPH_CAPABILITY_NO_EVIDENCE: &str = "no-evidence";
pub const ACTIVITY_MEMORY_GRAPH_INDEX_VERSION: &str = "activity-memory-v1";
pub const ACTIVITY_MEMORY_GRAPH_GAME_LABEL_HINT: &str = "game";
pub const ACTIVITY_MEMORY_GRAPH_REASON_EDGE_LIMIT: &str = "memory-edge-limit";

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActivityMemoryGraphNodeKind {
    #[serde(rename = "child-profile")]
    ChildProfile,
    #[serde(rename = "device")]
    Device,
    #[serde(rename = "browser-url")]
    BrowserUrl,
    #[serde(rename = "domain")]
    Domain,
    #[serde(rename = "video")]
    Video,
    #[serde(rename = "app")]
    App,
    #[serde(rename = "game")]
    Game,
    #[serde(rename = "activity-session")]
    ActivitySession,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActivityMemoryGraphEdgeKind {
    #[serde(rename = "visited")]
    Visited,
    #[serde(rename = "watched")]
    Watched,
    #[serde(rename = "played")]
    Played,
    #[serde(rename = "active-during")]
    ActiveDuring,
    #[serde(rename = "performed-by-child")]
    PerformedByChild,
    #[serde(rename = "derived-from-evidence")]
    DerivedFromEvidence,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActivityMemoryGraphEntryStatus {
    #[serde(rename = "usable")]
    Usable,
    #[serde(rename = "degraded")]
    Degraded,
    #[serde(rename = "stale")]
    Stale,
    #[serde(rename = "rejected")]
    Rejected,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActivityMemoryGraphQueryKind {
    #[serde(rename = "visited-urls")]
    VisitedUrls,
    #[serde(rename = "played-games")]
    PlayedGames,
    #[serde(rename = "watched-videos")]
    WatchedVideos,
    #[serde(rename = "activity-by-time-range")]
    ActivityByTimeRange,
    #[serde(rename = "explain-evidence")]
    ExplainEvidence,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityMemoryGraphTrace {
    pub entry_status: ActivityMemoryGraphEntryStatus,
    pub source_evidence_references: Vec<ParentEvidenceReference>,
    pub source_policy_version: Option<String>,
    pub source_parent_action_references: Vec<ParentActionReference>,
    pub generated_at: String,
    pub expires_at: Option<String>,
    pub confidence: f64,
    pub derived_index_version: String,
    pub degraded_reasons: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityMemoryGraphTimeRange {
    pub observed_from: String,
    pub observed_until: String,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityMemoryGraphNode {
    pub graph_id: String,
    pub node_id: String,
    pub node_kind: ActivityMemoryGraphNodeKind,
    pub label: String,
    pub child_profile: Option<ChildProfileReference>,
    pub device: Option<ParentDeviceReference>,
    pub trace: ActivityMemoryGraphTrace,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityMemoryGraphEdge {
    pub graph_id: String,
    pub edge_id: String,
    pub edge_kind: ActivityMemoryGraphEdgeKind,
    pub from_node_id: String,
    pub to_node_id: String,
    pub observed_from: String,
    pub observed_until: Option<String>,
    pub duration_ms: Option<u64>,
    pub trace: ActivityMemoryGraphTrace,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityMemoryGraphQuery {
    pub query_id: String,
    pub query_kind: ActivityMemoryGraphQueryKind,
    pub child_profile: Option<ChildProfileReference>,
    pub device: ParentDeviceReference,
    pub time_range: ActivityMemoryGraphTimeRange,
    pub as_of: String,
    pub limit: u64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityMemoryGraphReadModel {
    pub schema_version: u16,
    pub generated_at: String,
    pub custody: String,
    pub capability_status: String,
    pub query: ActivityMemoryGraphQuery,
    pub read_at: String,
    pub nodes: Vec<ActivityMemoryGraphNode>,
    pub edges: Vec<ActivityMemoryGraphEdge>,
    pub returned_node_count: u64,
    pub returned_edge_count: u64,
    pub omitted_edge_count: u64,
    pub degraded_reasons: Vec<String>,
}
